common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_message.h

```diff
@@ -67,51 +67,71 @@ class message_com {
-                    else
-                        action = beerocks_message::eAction::ACTION_NONE;
-                    if (action != beerocks_message::eAction::ACTION_NONE) {
-                        p_action_header->action()    = (beerocks_message::eAction)action;
-                        p_action_header->action_op() = action_op;
-                        p_action_header->id()        = id;
-                        return p_class;
-                    } else {
-                        // LOG(ERROR) << "action != beerocks_message::eAction::ACTION_NONE -> This shouldn't happen";
-                    }
-                }
-            }
-        } else {
-            std::cout << "beerocks_message.h[ " << __LINE__ << "]: "
-                      << "create vs message failed!" << std::endl;
+
+        auto actionhdr = cmdu_tx.addClass<beerocks_message::cACTION_HEADER>();
+        if (!actionhdr) {
+            std::cout << "beerocks_message.h[ " << __LINE__ << "]: " << __FUNCTION__ << " failed!" << std::endl;
+            return nullptr;
         }
+
+        p_class = cmdu_tx.addClass<T>();
+        if (!p_class) {
+            std::cout << "beerocks_message.h[ " << __LINE__ << "]: " << __FUNCTION__ << " failed!" << std::endl;
+            return nullptr;
+        }
+
+        auto action_op = p_class->get_action_op();
+        if (typeid(action_op) == typeid(beerocks_message::eActionOp_CONTROL))
+            action = beerocks_message::eAction::ACTION_CONTROL;
+        else if (typeid(action_op) == typeid(beerocks_message::eActionOp_PLATFORM))
+            action = beerocks_message::eAction::ACTION_PLATFORM;
+        else if (typeid(action_op) == typeid(beerocks_message::eActionOp_BACKHAUL))
+            action = beerocks_message::eAction::ACTION_BACKHAUL;
+        else if (typeid(action_op) == typeid(beerocks_message::eActionOp_APMANAGER))
+            action = beerocks_message::eAction::ACTION_APMANAGER;
+        else if (typeid(action_op) == typeid(beerocks_message::eActionOp_MONITOR))
+            action = beerocks_message::eAction::ACTION_MONITOR;
+        else if (typeid(action_op) == typeid(beerocks_message::eActionOp_CLI))
+            action = beerocks_message::eAction::ACTION_CLI;
+        else if (typeid(action_op) == typeid(beerocks_message::eActionOp_BML))
+            action = beerocks_message::eAction::ACTION_BML;
+        else
+            action = beerocks_message::eAction::ACTION_NONE;
+
+        if (action == beerocks_message::eAction::ACTION_NONE) {
+            std::cout << "beerocks_message.h[ " << __LINE__ << "]: " << __FUNCTION__ << " failed!" << std::endl;
+            return nullptr;
+        }
+
+        actionhdr->action()    = (beerocks_message::eAction)action;
+        actionhdr->action_op() = action_op;
+        actionhdr->id()        = id;
+
         return p_class;
     }
 
+    template <class T>
+    static std::shared_ptr<T> create_vs_message(ieee1905_1::CmduMessageTx &cmdu_tx, uint16_t id = 0)
```

Comment:
Suggestion:
`create_vs_message` should only create the VS CMDU,
After that we should call for `add_intel_vs_data<T>()` separately

Reply:
Agree, but its a huge change which I didn't want to make now just to add support for other vendors to use `create_vs_message` - not our focus.